src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -51,6 +52,36 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
             return default;
         }
 
+        /// <summary>Returns the element at a specified index in a sequence.</summary>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <param name="index">The index of the element to retrieve, which is either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="index" /> is outside the bounds of the <paramref name="source" /> sequence.
+        /// </exception>
+        /// <returns>The element at the specified position in the <paramref name="source" /> sequence.</returns>
+        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (!index.IsFromEnd)
+            {
+                return source.ElementAt(index.Value);
+            }
+
+            if (!TryGetElementAt(source, index.Value, out TSource? result))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index);
+            }
+
+            return result!;
```

Comment:
```suggestion
            return result;
```

Reply:
Strangely, if removing `!`, `dotnet build` works fine with 0 error and 0 warning, but in VS 2019 it always shows an error:
![image](https://user-images.githubusercontent.com/1563770/107312718-e4de6580-6a45-11eb-81e1-44e7570ca354.png)

@stephentoub Is this a VS problem? Should I keep `!`?

Reply:
What version of VS are you using?

Reply:
I am using VS 2019 16.8.5. I launched VS with `build.cmd -vs System.Linq`.

> Microsoft Visual Studio Enterprise 2019
> Version 16.8.5
> VisualStudio.16.Release/16.8.5+31005.135
> Microsoft .NET Framework
> Version 4.8.04084
> 
> Installed Version: Enterprise
> 
> Visual C++ 2019   00435-60000-00000-AA259
> Microsoft Visual C++ 2019
> 
> ADL Tools Service Provider   1.0
> This package contains services used by Data Lake tools
> 
> ASA Service Provider   1.0
> 
> ASP.NET and Web Tools 2019   16.8.559.8768
> ASP.NET and Web Tools 2019
> 
> ASP.NET Core Razor Language Services   16.1.0.2052803+84e121f1403378489b842e1797df2f3f5a49ac3c
> Provides languages services for ASP.NET Core Razor.
> 
> ASP.NET Web Frameworks and Tools 2019   16.8.559.8768
> For additional information, visit https://www.asp.net/
> 
> Azure App Service Tools v3.0.0   16.8.559.8768
> Azure App Service Tools v3.0.0
> 
> Azure Data Lake Node   1.0
> This package contains the Data Lake integration nodes for Server Explorer.
> 
> Azure Data Lake Tools for Visual Studio   2.6.3000.0
> Microsoft Azure Data Lake Tools for Visual Studio
> 
> Azure Functions and Web Jobs Tools   16.8.559.8768
> Azure Functions and Web Jobs Tools
> 
> Azure Stream Analytics Tools for Visual Studio   2.6.3000.0
> Microsoft Azure Stream Analytics Tools for Visual Studio
> 
> C# Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
> C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.
> 
> Common Azure Tools   1.10
> Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.
> 
> Cookiecutter   16.8.20241.2
> Provides tools for finding, instantiating and customizing templates in cookiecutter format.
> 
> Extensibility Message Bus   1.2.6 (master@34d6af2)
> Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.
> 
> Fabric.DiagnosticEvents   1.0
> Fabric Diagnostic Events
> 
> GhostDoc   20.2.20270.0
> Generate XML Comments from your code, maintain clean and up-to-date documentation, produce help documentation in multiple formats, use intelligent source code Spell Checker in Visual Studio.
> 
> GitHub.VisualStudio   2.11.106.19330
> A Visual Studio Extension that brings the GitHub Flow into Visual Studio.
> 
> IntelliCode Extension   1.0
> IntelliCode Visual Studio Extension Detailed Info
> 
> JetBrains ReSharper 2020.3.2   Build 203.0.20201229.111204
> JetBrains ReSharper package for Microsoft Visual Studio. For more information about ReSharper, visit http://www.jetbrains.com/resharper. Copyright © 2021 JetBrains, Inc.
> 
> Microsoft Azure HDInsight Azure Node   2.6.3000.0
> HDInsight Node under Azure Node
> 
> Microsoft Azure Hive Query Language Service   2.6.3000.0
> Language service for Hive query
> 
> Microsoft Azure Service Fabric Tools for Visual Studio   16.0
> Microsoft Azure Service Fabric Tools for Visual Studio
> 
> Microsoft Azure Stream Analytics Language Service   2.6.3000.0
> Language service for Azure Stream Analytics
> 
> Microsoft Azure Stream Analytics Node   1.0
> Azure Stream Analytics Node under Azure Node
> 
> Microsoft Azure Tools   2.9
> Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30924.1
> 
> Microsoft Continuous Delivery Tools for Visual Studio   0.4
> Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.
> 
> Microsoft JVM Debugger   1.0
> Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines
> 
> Microsoft Library Manager   2.1.113+g422d40002e.RR
> Install client-side libraries easily to any web project
> 
> Microsoft MI-Based Debugger   1.0
> Provides support for connecting Visual Studio to MI compatible debuggers
> 
> Microsoft Visual C++ Wizards   1.0
> Microsoft Visual C++ Wizards
> 
> Microsoft Visual Studio Tools for Containers   1.1
> Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.
> 
> Microsoft Visual Studio VC Package   1.0
> Microsoft Visual Studio VC Package
> 
> Mono Debugging for Visual Studio   16.8.43 (00471f8)
> Support for debugging Mono processes with Visual Studio.
> 
> Node.js Tools   1.5.20902.1 Commit Hash:b474efcb6f92db52a8f8e2e6a8cb9648476885cc
> Adds support for developing and debugging Node.js apps in Visual Studio
> 
> NuGet Package Manager   5.8.1
> NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/
> 
> ProjectServicesPackage Extension   1.0
> ProjectServicesPackage Visual Studio Extension Detailed Info
> 
> Python   16.8.20241.2
> Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.
> 
> Python - IronPython support   16.8.20241.2
> Provides templates and integration for IronPython-based projects.
> 
> Python - Profiling support   16.8.20241.2
> Profiling support for Python projects.
> 
> Snapshot Debugging Extension   1.0
> Snapshot Debugging Visual Studio Extension Detailed Info
> 
> SQL Server Data Tools   16.0.62102.01130
> Microsoft SQL Server Data Tools
> 
> SQLite & SQL Server Compact Toolbox   4.7
> SQLite & SQL Server Compact Toolbox adds scripting, import, export, rename, query execution and much more to SQL Server Compact & SQLite Data Connections.
> 
> Test Adapter for Boost.Test   1.0
> Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.
> 
> Test Adapter for Google Test   1.0
> Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.
> 
> ToolWindowHostedEditor   1.0
> Hosting json editor into a tool window
> 
> TypeScript Tools   16.0.21016.2001
> TypeScript Tools for Microsoft Visual Studio
> 
> Visual Basic Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
> Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.
> 
> Visual C++ for Linux Development   1.0.9.30608
> Visual C++ for Linux Development
> 
> Visual F# Tools   16.8.0-beta.20507.4+da6be68280c89131cdba2045525b80890401defd
> Microsoft Visual F# Tools
> 
> Visual Studio Code Debug Adapter Host Package   1.0
> Interop layer for hosting Visual Studio Code debug adapters in Visual Studio
> 
> Visual Studio Container Tools Extensions   1.0
> View, manage, and diagnose containers within Visual Studio.
> 
> Visual Studio Tools for CMake   1.0
> Visual Studio Tools for CMake
> 
> Visual Studio Tools for Containers   1.0
> Visual Studio Tools for Containers
> 
> Visual Studio Tools for Kubernetes   1.0
> Visual Studio Tools for Kubernetes
> 
> Visual Studio Tools for Unity   4.8.2.0
> Visual Studio Tools for Unity
> 
> VisualStudio.DeviceLog   1.0
> Information about my package
> 
> VisualStudio.Foo   1.0
> Information about my package
> 
> VisualStudio.Mac   1.0
> Mac Extension for Visual Studio
> 
> Xamarin   16.8.000.262 (d16-8@4d60f9c)
> Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.
> 
> Xamarin Designer   16.8.0.507 (remotes/origin/d16-8@e87b24884)
> Visual Studio extension to enable Xamarin Designer tools in Visual Studio.
> 
> Xamarin Templates   16.8.112 (86385a3)
> Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.
> 
> Xamarin.Android SDK   11.1.0.26 (d16-8/a36ce73)
> Xamarin.Android Reference Assemblies and MSBuild support.
>     Mono: 5e9cb6d
>     Java.Interop: xamarin/java.interop/d16-8@79d9533
>     ProGuard: Guardsquare/proguard/proguard6.2.2@ebe9000
>     SQLite: xamarin/sqlite/3.32.1@1a3276b
>     Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-8@2fb1cbc
> 
> 
> Xamarin.iOS and Xamarin.Mac SDK   14.10.0.4 (5a05865f6)
> Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

Reply:
Yeah, I think that version of the C# tools lacks some of the improvements around nullable analysis.  I have a newer version installed and don't see any warnings in VS.